### ISSUE TITLE
大模型原理

### ISSUE THREAD

**Issue by inkjun-com** [inkjun-com/note#2]
## Anaconda

Python 和 R 语言，专注于数据科学和机器学习领域的开发环境管理

Jupyter Notebook

pip 是 conda 共存但定位不同

conda 是跨语言的包和环境管理器

pip 的可执行环境：env/your_env_name/bin/pip

pip install -i 镜像源

https://www.anaconda.com/download/success

词向量，每一个词用一个向量表示

语言的概率，是根据前一个词预测下一个词的概率

神经概率模型如何去计算一个句子的规律

分词

定义一个窗口，窗口内滑动，构建训练数据

thou shalt not make a machine in the likeness of human mind

前向传播
1. 输入层：分词、词向量（随机初始化为一个 N 维向量）
2. 特征抽取：$y = b + Wx + Utanh(d + Hx)$
3. 输出层：接一个 softmax 函数，预测出下一个词是目标词的概率 `P(wt|wt-1)`

损失函数

后向传播更新参数

(b, d, W, U, H, C)

C 是我们之前随机初始化的向量，但之后会越来准确

词向量的价值

问题：无法处理一词多义的问题

## 词向量模型实战

https://modelscope.cn/models/NeuML/glove-6B/files

```python
# 模型下载
from modelscope import snapshot_download
model_dir = snapshot_download('NeuML/glove-6B', cache_dir='./')

from staticvectors import StaticVectors
model = StaticVectors("neuml/glove-6B/model.sqlite")
vecs = model.embeddings(['king', 'queen'])
print(vecs)

# 计算余弦相似度 a*b/|a||b|
import numpy as np
cos_sim = np.dot(vecs[0], vecs[1])/(np.linalg.norm(vecs[0]) * np.linalg.norm(vecs[1]))
print(cos_sim)

# 计算欧氏距离 |a-b|
dis = np.linalg.norm(vecs[0] - vecs[1])
print(dis)

# 可视化
from sklearn.manifold import TSNE
import warnings

word_list = ['boy', 'girl', 'woman', 'man', 'apple', 'banana', 'peach', 'purple']

# 若有无效值，替换为有效值
vecs = model.embeddings(word_list)

# 调整 perplexity 参数（小于样本数8）
# 使用上下文管理器仅在 TSNE 计算时抑制警告
with warnings.catch_warnings():
    warnings.filterwarnings("ignore", category=RuntimeWarning)
    tsne = TSNE(n_components=2, random_state=42, perplexity=2)
    vecs_2d = tsne.fit_transform(vecs)

print(vecs_2d)

import matplotlib.pyplot as plt

plt.rcParams["font.family"] = ["Heiti TC", "sans-serif"] # 设置中文字体
plt.rcParams["axes.unicode_minus"] = False  # 解决负号显示问题

plt.figure(figsize=(12, 8))
plt.scatter(vecs_2d[:, 0], vecs_2d[:, 1])

for i, word in enumerate(word_list):
    plt.annotate(word, (vecs_2d[i, 0], vecs_2d[i, 1]), fontsize=8)

plt.title('词向量可视化 TSNE 降维')
plt.show()
```

错误原因总结

tokenizers构建失败 ：

- 主要原因：Python 3.13.5存在包兼容性问题，且缺少必要的构建工具。
- 解决方案：安装Xcode Command Line Tools和Rust编译器解决构建依赖。

sentence_similarity安装失败 ：

- 主要原因：依赖项sentencepiece和tokenizers的构建问题。
- 解决方案：通过conda安装sentencepiece和tokenizers，然后使用 pip install --no-deps sentence_similarity 避免依赖冲突。

```shell
conda install -c conda-forge sentencepiece tokenizers -y
pip install --no-deps sentence_similarity modelscope
```

## 预训练模型实战

计算句子的相似度

```python
from modelscope import snapshot_download
snapshot_download('BAAI/bge-base-zh', cache_dir='./')

from sentence_transformers import SentenceTransformer
model = SentenceTransformer('BAAI/bge-base-zh')

sentence1 = "印刻君是一位前端程序员，分享有温度的轻知识，有深度的硬内容"
sentence2 = "白日依山尽，黄河入海流"

embedding1 = model.encode(sentence1)
embedding2 = model.encode(sentence2)

# 计算两个句子的相似度
similarity = embedding1@embedding2.T
print(similarity)

query = "印刻君是谁？"
query_embedding = model.encode(query)

similarity1 = query_embedding@embedding1.T
similarity2 = query_embedding@embedding2.T

print("{}, {}".format(similarity1, similarity2))
```

## 分词、tokenizer

```python
from transformers import AutoTokenizer

# 加载
# encode
# decode

tokenizer = AutoTokenizer.from_pretrained("./DeepSeek-V3")

result = tokenizer.encode("Hello Wrold")

print(result)

decodeResult = tokenizer.decode(result)

print(decodeResult)
```

### 训练过程

主体思想：统计相邻子词的频率，不断融合频率最高的相邻子词语

```
BPE 算法训练过程
输入：语料库（Corpus） C，词汇表最终尺寸 N
输出：最终尺寸为 N 的词汇表 V（排序过的），合并规则 M

BFE(C, N)
L <- 将 C 分割成一个最小单位标记（字符或字节）序列
V <- L 中的所有唯一字符（或 unicode 字节）

while |V| < N do # 合并 token
  <tL, tR> <- L 中最常见的重复字母
  tnew <- tL + tR # 连接成一个新的 token
  V <- V + [tnew] # 更新词汇表
  M <- M U {<tL, tR> -> tnew} # 更新合并规则
  L 替换所有 <tL, tR>, tnew
end while

return V, M
```

#### 词典构造例子

训练一个词典大小为 8 的字典，训练语料为：abcccabcabdef

字符串包含 a、b、c、d、e、f，首先分给它们一个 id，作为字典的第一个版本：

```json
{ "a": 0, "b": 1, "c": 2, "d": 3, "e": 4, "f": 5 }
```

原始的输入变成 id：

```json
[0, 1, 2, 2, 2, 0, 1, 2, 0, 1, 3, 4, 5]
```

统计相邻的 id 对出现的频率，得到

```json
{  "(a, b)": 3, "(b, c)": 4, "(c, c)": 2, "(c, a)": 2, "(b, d)": 1, "(d, e)": 1, "(e, f)": 1}
```

可以看到，"(a, b)" 这个子对出现的频率最高。我们把它们合并起来并在词典里面加入一个新的 id，并记录 merge 的过程

```json
{ "0": "a", "1": "b", "2": "c", "3": "d", "4": "e", "5": "f", "6": "ab" }
```

原始输入的 id 里面也把对应的 (0, 1) 替换成 6，也就是把 (a, b) 替换成 ab。替换后，原始输入 id 从 `[0, 1, 2, 2, 2, 0, 1, 2, 0, 1, 3, 4, 5]` 变为 `[6, 2, 2, 2, 6, 2, 6, 3, 4, 5]`

继续统计替换后的相邻 id 出现的频率，得到：

```json
{ "(ab, c)": 2, "(c, c)": 2, "(c, ab)": 2, "(ab, d)": 1, "(d, e)": 1, "(e, f)": 1 }
```

里面有三个 id 对出现的频率都是 2，任何选择一个进行融合。这里选择第一个，也就是 (ab, c) 进行融合，并记录 merge 的过程为 (6, 2) -> 7，融合后的词表为：

```json
{ "0": "a", "1": "b", "2": "c", "3": "d", "4": "e", "5": "f", "6": "ab", "7": "abc" }
```

原始输入变为：

```json
[7, 2, 2, 7, 6, 3, 4, 5]
```

继续统计相邻 id 的频率，得到

```json
{ "(abc, c)": 1, "(c, c)": 1, "(c, abc)": 1, "(abc, ab)": 1, "(ab, d)": 1, "(d, e)": 1, "(e, f)": 1 }
```

继续融合第一个最大的，得到新的字典，记录 merge 的过程为 (7, 2) -> 8

```json
{ "0": "a", "1": "b", "2": "c", "3": "d", "4": "e", "5": "f", "6": "ab", "7": "abc", "8": "abcc" }
```

词典的大小已经达到我们设定的大小，训练完成

#### 词典编码过程

假设输入的字符床 abc

先把单个字符转换成对应的 id：a -> 0, b -> 1, c -> 2，得到：`[0, 1, 2]`

对 `[0, 1, 2]` 进行 merge 动作，得到 [6, 2]，**每一步都选择可以 merge 的最小 id**，继续这个过程，最后得到编码结果为 [7]

```python
# 1. 导入或定义 BFETokenizer 类
class BFETokenizer:
    def __init__(self):
        self.merge = {}
        self.id_2_char = {}
        self.char_2_id = {}

    def train(self, input_texts, vocab_size):
        # 1. 对输入语料进行拆分
        unique_chars = sorted(list(set(input_texts)))

        # 2. 得到一个初始化词典
        id_2_char = {idx: char for idx, char in enumerate(unique_chars)}

        char_2_id = {char: idx for idx, char in enumerate(unique_chars)}

        # 3. 对输入语料进行 id 化
        ids = [char_2_id[c] for c in input_texts]
        print(f"ids: {ids}")

        '''
        这行代码的作用是 计算需要执行的合并操作次数
        - vocab_size ：用户指定的 目标词汇表大小 （函数参数）
        - len(unique_chars) ： 初始词汇表大小 （即输入文本中唯一字符的数量）
        - 两者的差值就是需要合并的次数
        例如：如果输入文本有 50 个唯一字符，目标词汇表大小是 100，那么需要执行 50 次合并操作（每次合并会产生一个新的子词，从而增加词汇表大小）。
        '''
        merge_times = vocab_size - len(unique_chars)

        '''
        这行代码的作用是 重新初始化当前词汇表大小计数器 。
        - 这里的 vocab_size 从原来的目标大小被重新赋值为 初始唯一字符数减一
        - 这是因为在后续的合并循环中，每次合并后会将 vocab_size += 1 （见原代码），这样可以确保最终词汇表大小能准确达到目标值
        '''
        vocab_size = len(unique_chars) - 1
        merge = {}

        # 4. 训练，合并子词，直到字典的大小达到目标值
        for i in range(merge_times):
            if len(ids) == 1:
                break

            # 统计相邻子词出现的频率
            stats = self.stats(ids)

            # 找出出现频率最高的相邻字符对
            pair = max(stats, key=stats.get)
            vocab_size += 1
            id_2_char[vocab_size] = id_2_char[pair[0]] + id_2_char[pair[1]]
            char_2_id[id_2_char[pair[0]] + id_2_char[pair[1]]] = vocab_size
            merge[pair] = vocab_size

            # 根据当前的词典，合并 ids
            ids = self.merge_ids(ids, pair, vocab_size)

        self.merge = merge
        self.char_2_id = char_2_id
        self.id_2_char = id_2_char

    def stats(self, ids):
        count = {}
        for item in zip(ids[:-1], ids[1:]):
            count[item] = count.get(item, 0) + 1
        return count

    def merge_ids(self, ids, pair, idx):
        new_ids = []
        i = 0
        while i < len(ids):
            if ids[i] == pair[0] and i < len(ids) - 1 and ids[i + 1] == pair[1]:
                new_ids.append(ids[i])
                i += 2
            else:
                new_ids.append(ids[i])
                i += 1
        return new_ids

    # 添加一个分词方法以便使用
    def tokenize(self, text):
        # 将文本转换为ID序列
        ids = [self.char_2_id.get(c, -1) for c in text]  # 处理未见过的字符
        ids = [id for id in ids if id != -1]  # 过滤掉未见过的字符

        # 应用合并规则
        for pair, new_id in self.merge.items():
            i = 0
            new_ids = []
            while i < len(ids):
                if i < len(ids) - 1 and ids[i] == pair[0] and ids[i + 1] == pair[1]:
                    new_ids.append(new_id)
                    i += 2
                else:
                    new_ids.append(ids[i])
                    i += 1
            ids = new_ids

        # 将ID转换回子词
        tokens = [self.id_2_char[id] for id in ids]
        return tokens


# 2. 创建分词器实例
tokenizer = BFETokenizer()

# 3. 准备训练数据
corpus = "abcccabcabdef"

# 4. 训练分词器
vocab_size = 8  # 目标词典大小
tokenizer.train(corpus, vocab_size)
```

### 推理过程

Algorathm BPE Inference

Input:
Output:

procedure BPE
